bpf-prog/block-iptables/bpf/src/block_iptables.bpf.c

```diff
@@ -0,0 +1,181 @@
+
+    // Safely get parent task_struct
+    parent_task = BPF_CORE_READ(task, real_parent);
+    if (!parent_task)
+        return 0;
+
+    // Safely read parent->comm
+    if (bpf_core_read_str(&parent_comm, sizeof(parent_comm), &parent_task->comm) < 0)
+        return 0;
+
+    // Check if parent_comm is an allowed command
+    #pragma unroll
+    for(int p = 0; p < COMM_COUNT; p++) {
+        int match = 1;
+        for (int i = 0; i < TASK_COMM_LEN && target_prefixes[p][i] != '\0'; i++) {
+            if (parent_comm[i] != target_prefixes[p][i]) {
+                match = 0;
+                break;
+            }
+        }
+
+        if(match) {
+            bpf_printk("Allowed netlink from parent: %s\n", parent_comm);
+            return 1;
+        }
+    }
+
+
+    return 0; // Block
+}
+
+int is_host_ns() {
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    struct nsproxy *nsproxy;
+    struct net *net_ns;
+    u32 netns_ino = 0;
+
+    nsproxy = BPF_CORE_READ(task, nsproxy);
+    if (!nsproxy)
+        return 0;
+
+    net_ns = BPF_CORE_READ(nsproxy, net_ns);
+    if (!net_ns)
+        return 0;
+
+    netns_ino = BPF_CORE_READ(net_ns, ns.inum);
+
+    if (netns_ino != host_netns_inode) {
+        return 0;
+    }
+
+    return 1;
+}
+
+void increment_event_counter() {
+    u32 key = 0;
+    u64 *value;
+
+    value = bpf_map_lookup_elem(&event_counter, &key);
+    if (value) {
+        __sync_fetch_and_add(value, 1);
+    } else {
+        u64 initial_value = 1;
+        bpf_map_update_elem(&event_counter, &key, &initial_value, BPF_ANY);
+    }
+}
+
+SEC("lsm/socket_setsockopt")
+int BPF_PROG(iptables_legacy_block, struct socket *sock, int level, int optname)
+{
+    if (sock == NULL) {
+        return 0;
+    }
+
+    if (level == 0 /*IPPROTO_IP*/ || level == 41 /*IPPROTO_IP6*/) {
+        if (optname == IPT_SO_SET_REPLACE) { 
+            if (is_host_ns() && !is_allowed_parent()) {
+                increment_event_counter();
+                return -EPERM;
+            }
+        }
+    }
+
+    return 0;
+}
+
+SEC("lsm/netlink_send")
+int BPF_PROG(iptables_nftables_block, struct sock *sk, struct sk_buff *skb) {
+    __u16 family = 0, proto = 0;
+    if (sk != NULL) {
+        bpf_probe_read_kernel(&family, sizeof(family), &sk->sk_family);
+    }
+
+    if (family != 16) // Not AF_NETLINK
+        return 0;
+
+    if (sk != NULL) {
+        bpf_probe_read_kernel(&proto, sizeof(proto), &sk->sk_protocol);
+    }
+
+    if (proto != 12) // Not NETLINK_NETFILTER
+        return 0;
+
+    if (!is_host_ns()) {
+        return 0;
+    }
+
+    struct nlmsghdr nlh = {};
+    void *data = NULL;
+    __u32 skb_len = 0;
+
+    if (bpf_core_read(&data, sizeof(data), &skb->data) < 0)
+        return 0;
+
+    if (!data)
+        return 0;
+
+    if (bpf_core_read(&skb_len, sizeof(skb_len), &skb->len) < 0)
+        return 0;
+
+    #pragma unroll
+    for (int i = 0; i < 4; i++) {
```

Comment:
is this like looking at 4 netlink messages in the buffer at a time? Or what is the reason for this number?

Reply:
BPF requires loops to be bounded, so we need to have some limit. Also, it's not really 4 messages at a time, it's 4 messages overall. In the experiments I've tried with iptables, the rule add message seems to be the second one always (after a "batch begin"). We could raise it or lower it, but we definitely need a limit.